src/Interpreters/TreeOptimizer.cpp

```diff
@@ -566,6 +581,127 @@ void transformIfStringsIntoEnum(ASTPtr & query)
+        return false;
+    auto predict_func = predicate->as<ASTFunction>();
+    if (! predict_func || predict_func->name != "equals")
+        return false;
+    auto ast_predict_column = predict_func->arguments->children[0]->as<ASTIdentifier>();
+    auto ast_predict_value = predict_func->arguments->children[1]->as<ASTLiteral>();
+    if (! ast_predict_column || ! ast_predict_value)
+        return false;
+    String predict_column = ast_predict_column->name();
+    Field predict_value = ast_predict_value->value;
+
+    // Parse partition_by in create query and get partition id to apply in ALTER DROP PARTITION.
+    StorageID table_id(*dynamic_pointer_cast<ASTQueryWithTableAndOutput>(query));
+    if (table_id.database_name.empty()) 
+        table_id.database_name = context.getCurrentDatabase();
+    auto db_tbl = DatabaseCatalog::instance().getDatabaseAndTable(table_id, context);
+    auto database = db_tbl.first;
+    auto table = db_tbl.second;
+    ASTPtr create_ptr = database->getCreateTableQuery(table_id.getTableName(), context);
+    auto create_query = create_ptr->as<ASTCreateQuery>();
+    if (! create_query || ! create_query->storage)
+        return false;
+    auto partition_by = create_query->storage->partition_by;
+    if (auto partition_by1 = partition_by->as<ASTIdentifier>())
+    {
+        // Partition key must be equal to predict_column
+        if (partition_by1->name() != predict_column)
+            return false;
+        partition = predict_value;
+        return true;
+    }
+
+    if (auto partition_by2 = partition_by->as<ASTFunction>())
+    {
+        // Partition key must be equal to predict_column
+        const auto& children = partition_by2->arguments->children;
+        if (children.size() != 1 || ! children[0]->as<ASTIdentifier>() ||
+            children[0]->as<ASTIdentifier>()->name() != predict_column)
+            return false;
+
+        const auto& func_name = partition_by2->name;
+        auto metadata = table->getInMemoryMetadataPtr();
+        const auto& col_desc = metadata->getColumns().get(predict_column);
+        if (predict_value.getType() == Field::Types::String) 
+        {
+            DataTypeDate::FieldType to;
+            const auto& tmp = predict_value.get<String>();
+            ReadBufferFromMemory read_buffer(tmp.c_str(), tmp.size());
+            parseImpl<DataTypeDate>(to, read_buffer, nullptr);
+            predict_value = to;
+        }
+        auto column = col_desc.type->createColumnConst(1, predict_value);
+        const ColumnWithTypeAndName arg{column, col_desc.type, predict_column};
+        const ColumnsWithTypeAndName args{std::move(arg)};
+        const FunctionFactory & func_factory = FunctionFactory::instance();
+        auto func_builder = func_factory.get(func_name, context);
+
+        // Funtion must satisfy one of those conditions:
+        // 1. Function must be injective 
+        // 2. Function is "toYYYYMMDD", and partition_key column has Date Type
```

Comment:
This case looks extremely specific. We can cover a much wider range of cases if we will perform this optimization inside MergeTree.

Reply:
Got it. I will try to make it more general.